### DISCUSSION TITLE
Convert ONNX models to TensorRT online

### DISCUSSION THREAD

**Discussion by madsciencetist** [blakeblackshear/frigate#6572] [Ideas]
TensortRT models are specific to both hardware and library versions, so generally speaking, they are not shareable. Even for a single user, the model-conversion advice [given in the docs](https://docs.frigate.video/configuration/detectors/#generate-models) does not scale to new versions, because the docker image does not necessarily match the runtime image.

I propose that the `model_path` used by the tensorrt detector plugin not be a `.trt` model file, but an `.onnx` file. The plugin itself would run the ONNX->TRT conversion and cache the output.

Benefits:
- TRT model generation is guaranteed to happen on the same platform & environment in which Frigate runs
- ONNX models can be shared between users with no extra preprocessing
- TRT options, like running on a DLA vs. GPU, can be specified via detector configs rather than being baked into a manually-generated  `.trt` file
  - I'm not sure what further Frigate changes would be necessary to support this, but it would simplify specifying multiple detectors on Jetson devices, e.g. one that uses DLA0, a second that uses DLA1, and a third that uses the GPU
- The YOLO->ONNX conversion is at least much easier because there is no TensorRT/NVIDIA dependency, and at best nonexistent because `.onnx` files can be released/downloaded directly

Cons:
- There will be a long (10s of seconds?) detector startup delay the first time Frigate is started after the model is changed, the DLA mode is changed, or the model cache is cleared

Checking for community agreement before implementing...thoughts?

**Comment by NickM-27** [comment]
This would require more changes from frigate including a way to configure which model is used. Also, it would need to handle cases where a user does want to provide their own model. Similarly if the first steps aren't needed then the user may not map the trt-models folder and that would lead to this work needing to be done on each restart.

**Reply by madsciencetist** [reply]
> This would require more changes from frigate including a way to configure which model is used

Could you elaborate? I know the "multiple detectors on Jetson devices" part would require changes, but let's set that aside for the near term, and for everything else I was envisioning code changes only within `tensorrt.py`. There is already configuration to specify that a detector is of `type: tensorrt`, and then within `tensorrt.py`, we'd invoke the TRT engine builder. Is your point that we don't have a way to pass parameters to the engine builder, e.g. to specify which DLA to use?

---

> it would need to handle cases where a user does want to provide their own model

The user could absolutely provide their own ONNX model, just as they can today provide their own TRT model. AFAIK every TRT model comes from an ONNX conversion. If we do want to retain the ability to specify a TRT model, e.g. for backwards compatibility, we could allow the `model_path` to point to either ` .trt` or `.onnx` file, and then convert or not based on file extension.

---

> if the first steps aren't needed then the user may not map the trt-models folder and that would lead to this work needing to be done on each restart.

I think this is just a question of where to cache the generated TRT models. Unless we supply the ONNX models, the user would still have to mount their ONNX model directory, so we could write the TRT models to the same directory. Or we could use `/tmp/cache`. Or we could add a new `/media/frigate/model_cache` directory that the user is encouraged to mount, but it would still function if they didn't. The instructions to mount the model_cache would be next to the instructions to mount the ONNX models in the docs, and if the user still doesn't, honestly I don't think they'll notice if it takes a minute for detections to start working after a reboot.

**Reply by NickM-27** [reply]
> Could you elaborate? I know the "multiple detectors on Jetson devices" part would require changes, but let's set that aside for the near term, and for everything else I was envisioning code changes only within tensorrt.py. There is already configuration to specify that a detector is of type: tensorrt, and then within tensorrt.py, we'd invoke the TRT engine builder. Is your point that we don't have a way to pass parameters to the engine builder, e.g. to specify which DLA to use?

The current flow allows a multitude of different models to be built like yolov4, yolov4-tiny, yolov7, yolov7-tiny, etc, etc. So there would need to be a config to be able to specify this. I don't think in the frigate container itself that an env variable is the right way to do this.

> If we do want to retain the ability to specify a TRT model, e.g. for backwards compatibility, we could allow the model_path to point to either  .trt or .onnx file, and then convert or not based on file extension.

This feature is in stable, backwards compatibility is a must all the way through IMO

> I think this is just a question of where to cache the generated TRT models. Unless we supply the ONNX models, the user would still have to mount their ONNX model directory, so we could write the TRT models to the same directory. Or we could use /tmp/cache. Or we could add a new /media/frigate/model_cache directory that the user is encouraged to mount, but it would still function if they didn't. The instructions to mount the model_cache would be next to the instructions to mount the ONNX models in the docs

With frigate+ there is already a /config/model_cache folder so that could work. writing to RAM will absolutely not work and is a bad practice in docker container design.

**Reply by madsciencetist** [reply]
> The current flow allows a multitude of different models to be built like yolov4, yolov4-tiny, yolov7, yolov7-tiny, etc, etc. So there would need to be a config to be able to specify this. I don't think in the frigate container itself that an env variable is the right way to do this.

Outside of Frigate, a script is currently provided to download and convert multiple YOLO models to TRT models. That script could remain, but be reduced to downloading and converting multiple YOLO models to ONNX models. The docker image would become much smaller and would no longer be platform-specific.

I think there's another feature that can be discussed to automatically supply those ONNX files, removing the necessity of the script entirely, but we can leave that aside for now.

> This feature is in stable, backwards compatibility is a must all the way through IMO

No problem.

> With frigate+ there is already a /config/model_cache folder

Perfect.

**Reply by madsciencetist** [reply]
> This would require more changes from frigate including a way to configure which model is used

In case I missed your point with this - the way to specify the model is with the existing `model:` parameter.

**Reply by NickM-27** [reply]
> > The current flow allows a multitude of different models to be built like yolov4, yolov4-tiny, yolov7, yolov7-tiny, etc, etc. So there would need to be a config to be able to specify this. I don't think in the frigate container itself that an env variable is the right way to do this.
> 
> Outside of Frigate, a script is currently provided to download and convert multiple YOLO models to TRT models. That script could remain, but be reduced to downloading and converting multiple YOLO models to ONNX models. The docker image would become much smaller and would no longer be platform-specific.
> 
> I think there's another feature that can be discussed to automatically supply those ONNX files, removing the necessity of the script entirely, but we can leave that aside for now.

I'm not sure I follow, what you're saying doesn't seem immediately relevant to the concern that I mentioned above

**Reply by NickM-27** [reply]
Nevermind I think I see what you mean, my concern would only be relevant to the later step where ONNX models are supplied automatically.

**Reply by NickM-27** [reply]
I'd also be curious to hear @NateMeyer opinion since the initial implementation was done by him

**Reply by blakeblackshear** [reply]
I think this would be great if you can figure out how to make this work. I would like to ultimately offer frigate+ models as onnx files as well.

A couple things:
- I want the model_hash to be calculated on the onnx file
- Make use of the model_cache folder so the model only has to be built once

**Reply by NateMeyer** [reply]
I think ONNX is a good distribution format, most libraries I've seen support converting them and it would be nice to support one model in the various detectors.  

For tensorRT, I'm not sure if we have all the libraries included to do the model building in the fridge container, I focused on getting only the necessary components of the runtime to keep the image size reasonable.  The same would go for OpenVino, and I don't know about the other detectors either.  My previous assumption was to keep libraries for generating/converting models in their own stage of the image build process, again so it didn't blast the final images.

**Comment by madsciencetist** [comment]
I'm backing off from this idea. A docker image for offline TRT model conversion is 11 GB and takes 10 minutes to convert yolov7 on a Xavier NX. I have trouble believing we can optimize it sufficiently to convert at runtime.
